File: compiler/nodes/expressions/string.go
package expressions

import (
	"strconv"

	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type StringExpressionNode struct {
	Value string
	Facts *database.Facts
}

func (node *StringExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseStringExpression(parser *syntax.Parser) (*StringExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	value, err := parser.Token("String")
	if err != nil {
		return nil, err
	}

	return &StringExpressionNode{
		Value: value,
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *StringExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	stringType := &types.NamedTypeNode{
		Name:       "String",
		Parameters: nil,
		Facts:      database.NewFacts(database.GetSpanFact(node)),
	}
	visitor.Visit(stringType)
	visitor.Db.Graph.Replace(stringType, nil)

	visitor.Constraint(typecheck.NewGroupConstraint(node, stringType))
}

func (node *StringExpressionNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)
	c.WriteString(span, strconv.Quote(node.Value))
	return nil
}

File: compiler/src/nodes/expressions/string.rs
use crate::{
    codegen::{Codegen, CodegenCtx, ir},
    database::{HiddenNode, Node, NodeRef},
    nodes::{NamedTypeNode, visit_expression},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct StringExpressionNode {
    pub value: String,
}

impl Node for StringExpressionNode {}

pub fn parse_string_expression(
    parser: &mut Parser<'_>,
) -> Result<StringExpressionNode, ParseError> {
    let value = parser.token(TokenKind::String)?;

    Ok(StringExpressionNode { value })
}

impl Visit for StringExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        let span = visitor.span(node);
        let string_type = visitor.node(
            span,
            HiddenNode::new(NamedTypeNode {
                name: String::from("String"),
                parameters: Vec::new(),
            }),
        );

        visitor.visit(&string_type);
        visitor.constraint(GroupConstraint::new(node.clone(), string_type));
    }
}

impl Codegen for StringExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> Option<ir::SpannedExpression> {
        ir::Expression::String(self.value.clone()).at(node, ctx)
    }
}
